orttraining/orttraining/core/graph/optimizer_graph_builder.cc

```diff
@@ -443,17 +466,26 @@ Status OptimizerGraphBuilder::BuildInternal(
   // check if all gradients are finite
   ArgDef global_grad_norm_argdef;
   ArgDef global_grad_norm_finite_argdef;
-  if (opt_graph_config_.use_mixed_precision) {
-    //gradient norm for bfloat is not ready yet. skip it to unblock the testing
+
+  if (opt_graph_config_.enable_grad_norm_clip ||
+      (opt_graph_config_.use_mixed_precision &&
+       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16)) {
+    //gradient norm for bfloat16 is not ready yet. skip it to unblock the testing
     //will add it back when it is ready
-    if (opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
-      ORT_RETURN_IF_ERROR(AddGradientNorm(
-          nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
-      ORT_RETURN_IF_ERROR(AddFiniteGradientCheck(
-          nodearg_name_generator, {global_grad_norm_argdef}, graph_defs, global_grad_norm_finite_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GradientAllIsFinite] = global_grad_norm_finite_argdef.name;
-    }
+    ORT_RETURN_IF_ERROR(AddGradientNorm(
+        nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
+    optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
+  }
+
+  if (opt_graph_config_.use_mixed_precision &&
+      opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
+    ORT_RETURN_IF_ERROR(AddFiniteGradientCheck(
+        nodearg_name_generator, {global_grad_norm_argdef}, graph_defs, global_grad_norm_finite_argdef));
+    optimizer_graph_outputs[OptimizerOutputKey::GradientAllIsFinite] = global_grad_norm_finite_argdef.name;
+  }
+
+  if (global_grad_norm_argdef.name.empty() && global_grad_norm_finite_argdef.name.empty()) {
```

Comment:
>global_grad_norm_argdef.name.empty() [](start = 6, length = 36)

Use `graph.GetNodeArg(global_grad_norm_argdef.name) != null` to determine if it exists.

Reply:
They are equivalent. 
At this point, then change is still in GraphDefs, and hasn't been applied on the graph.

Reply:
If there is an official way of checking if `ArgDef` exists, we should only use it instead of having several alternative ways (it's like Python v.s. Perl and the current winner is Python). After reading the document of `Graph`, I believe `GetNodeArg` is the right way.

---
In reply to: [520921889](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520921889) [](ancestors = 520921889)

Reply:
OK. I added a Exist() function to the ArgDef class.

Reply:
The signature of the new `Exist` better to be `Exist(const Graph& graph)` because graph is the scope of declaration. Inside the new `Exist`, there could be
```
bool ArgDef::ExistIn(const Graph& graph) {
  return graph.GetNodeArg(global_grad_norm_argdef.name) != null;
}
```

---
In reply to: [520966059](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520966059) [](ancestors = 520966059)

Reply:
discussed offline. the nodearg hasn't been added to graph at this point.